src/order/filter/filter_product.lean

```diff
@@ -0,0 +1,293 @@
+  { change _ ∈ _ at c', 
+    simp only [set.set_of_false, zero_ne_one, empty_in_sets_eq_bot] at c', 
+    exact NT c' },
+  ..filter_product.has_zero, 
+  ..filter_product.has_one }
+
+instance [division_ring β] (U : is_ultrafilter φ) : division_ring β* :=
+{ mul_inv_cancel := λ x, quotient.induction_on' x $ λ a hx, quotient.sound' $
+    have hx1 : _ := (not_imp_not.mpr quotient.eq'.mpr) hx,
+    have hx2 : _ := (ultrafilter_iff_compl_mem_iff_not_mem.mp U _).mpr hx1,
+    have h : {n : α | ¬a n = 0} ⊆ {n : α | a n * (a n)⁻¹ = 1} := 
+      by rw [set.set_of_subset_set_of]; exact λ n, division_ring.mul_inv_cancel,
+    mem_sets_of_superset hx2 h,
+  inv_mul_cancel := λ x, quotient.induction_on' x $ λ a hx, quotient.sound' $
+    have hx1 : _ := (not_imp_not.mpr quotient.eq'.mpr) hx,
+    have hx2 : _ := (ultrafilter_iff_compl_mem_iff_not_mem.mp U _).mpr hx1,
+    have h : {n : α | ¬a n = 0} ⊆ {n : α | (a n)⁻¹ * a n = 1} := 
+      by rw [set.set_of_subset_set_of]; exact λ n, division_ring.inv_mul_cancel,
+    mem_sets_of_superset hx2 h, 
+  ..filter_product.ring, 
+  ..filter_product.has_inv, 
+  ..filter_product.zero_ne_one_class U.1 }
+
+instance [field β] (U : is_ultrafilter φ) : field β* :=
+{ ..filter_product.comm_ring, 
+  ..filter_product.division_ring U }
+
+noncomputable instance [discrete_field β] (U : is_ultrafilter φ) : discrete_field β* :=
+{ inv_zero := quotient.sound' $ by show _ ∈ _;
+    simp only [inv_zero, eq_self_iff_true, (set.univ_def).symm, univ_sets],
+  has_decidable_eq := by apply_instance, 
+  ..filter_product.field U }
+
+instance [has_le β] : has_le β* := { le := lift_rel₂ has_le.le }
+
+instance [has_lt β] : has_lt β* := { lt := lift_rel₂ has_lt.lt }
+
+instance [preorder β] : preorder β* :=
+{ le_refl := λ x, quotient.induction_on' x $ λ a, show _ ∈ _, 
+    by simp only [le_refl, (set.univ_def).symm, univ_sets],
+  le_trans := λ x y z, quotient.induction_on₃' x y z $ λ a b c hab hbc, 
+    by filter_upwards [hab, hbc] λ i hiab hibc, le_trans hiab hibc, 
+  ..filter_product.has_le }
+
+instance [partial_order β] : partial_order β* :=
+{ le_antisymm := λ x y, quotient.induction_on₂' x y $ λ a b hab hba, quotient.sound' $ 
+    have hI : {n | a n = b n} = _ ∩ _ := set.ext (λ n, le_antisymm_iff), 
+    show _ ∈ _, by rw hI; exact inter_sets _ hab hba
+  ..filter_product.preorder }
+
+instance [linear_order β] (U : is_ultrafilter φ) : linear_order β* :=
+{ le_total := λ x y, quotient.induction_on₂' x y $ λ a b, 
+    have hS : _ ⊆ {i | b i ≤ a i} := λ i, le_of_not_le,
+    or.cases_on (mem_or_compl_mem_of_ultrafilter U {i | a i ≤ b i})
+      (λ h, or.inl h) 
+      (λ h, or.inr (sets_of_superset _ h hS))
+  ..filter_product.partial_order }
+
+end filter_product
+
+end filter
+
+namespace hyperreal
+
+open filter filter.filter_product
+
+/-- Hyperreal numbers on the ultrafilter extending the cofinite filter -/
+@[reducible] def hyperreal := filter.filterprod ℝ (@hyperfilter ℕ)
```

Comment:
move this definition before `namespace hyperreal` otherwise it is called `hyperreal.hyperreal` and the `r.const` as `hyperreal.const (r : hyperreal)` notation doesn't work.